csharp/src/Microsoft.ML.OnnxRuntime/InferenceSession.shared.cs

```diff
@@ -740,35 +885,35 @@ private static OrtValue ExtractOrtValueForOutput(NamedOnnxValue output, NodeMeta
         /// <param name="metaLookup">Metadata lookup function (input/initializers/output)</param>
         /// <param name="cleanupList">list to cleanup in an exception safe manner</param>
         /// <returns></returns>
-        private IntPtr[] GetOrtValuesHandles(IReadOnlyCollection<NamedOnnxValue> values, MetadataLookup metaLookup,
-            OrtValueExtractor ortValueExtractor,
-            DisposableList<IDisposable> cleanupList)
-        {
-            cleanupList.Capacity += values.Count * 2;
-            IntPtr[] result = new IntPtr[values.Count];
-            for (int valueIndex = 0; valueIndex < values.Count; ++valueIndex)
-            {
-                var value = values.ElementAt(valueIndex);
-                var meta = metaLookup(value.Name);
-                var ortValue = ortValueExtractor(value, meta, out IDisposable memHolder);
-                if (memHolder != null)
-                {
-                    cleanupList.Add(memHolder);
-                }
-                if (ortValue != null)
-                {
-                    cleanupList.Add(ortValue);
-                    result[valueIndex] = ortValue.Handle;
-                }
-                else
+        private static IntPtr[] GetOrtValuesHandles(IReadOnlyCollection<NamedOnnxValue> values, MetadataLookup metaLookup,
+            OrtValueHandleExtractor ortValueExtractor, out DisposableArray<IDisposable> disposer)
+        {
+            IDisposable[] memHolders = new IDisposable[values.Count];
+            var disp = new DisposableArray<IDisposable>(memHolders);
+            try
+            {
+                IntPtr[] result = new IntPtr[values.Count];
+                for (int valueIndex = 0; valueIndex < values.Count; ++valueIndex)
                 {
-                    result[valueIndex] = IntPtr.Zero;
+                    var value = values.ElementAt(valueIndex);
+                    var meta = metaLookup(value.Name);
+                    result[valueIndex] = ortValueExtractor(value, meta, out IDisposable memHolder);
+                    if (memHolder != null)
+                    {
+                        memHolders[valueIndex] = memHolder;
+                    }
                 }
+                disposer = disp;
+                return result;
+            }
+            catch (Exception)
+            {
+                disp.Dispose();
+                throw;
```

Comment:
Why `catch` here where elsewhere you're doing the disposal in `finally`?

Reply:
I will change it. Most of the time, I do not need disposing those on error.